### ISSUE TITLE
select one shot repositroy

### ISSUE THREAD

**Issue by frederikdammeier** [frederikdammeier/unipd-cvcs#2]
### RetinaNet (single-shot, anchor-based) from `torchvision`

**Summary**  
We can use `torchvision`’s `retinanet_resnet50_fpn_v2` as our **single-shot, anchor-based object detector**, pre-trained on **MS COCO**.  
This model will be one of the baselines in our comparison against other detectors (two-stage / anchor-free, etc.).

---

#### Why RetinaNet?

- **Single-stage (single-shot)**: detection in one pass → simpler and faster than two-stage models.
- **Anchor-based**: uses predefined anchors, which makes it directly comparable to Faster R-CNN / SSD.
- **Pre-trained on COCO**: no need to train from scratch; good performance out of the box.
- **Official `torchvision` implementation**: stable API, minimal setup, easy to integrate with our pipeline.

---

#### Subtasks (how to use)

- [ ] Add a small module/class that wraps RetinaNet inference (load model + transforms).
- [ ] Run RetinaNet on our COCO subset and save predictions to JSON/CSV.
- [ ] (Optional) Map COCO label IDs → class names for nicer logs/plots.

---

#### Documentation / References

**Official RetinaNet documentation in torchvision**
- https://docs.pytorch.org/vision/main/models/retinanet.html  
- https://docs.pytorch.org/vision/main/models/generated/torchvision.models.detection.retinanet_resnet50_fpn_v2.html  

**Tutorials / articles with RetinaNet + torchvision**
- https://stackabuse.com/retinanet-object-detection-with-pytorch-and-torchvision/  

**Optional: standalone RetinaNet PyTorch repository**
- https://github.com/yhenon/pytorch-retinanet  
  
You can extend this with:
- a helper function `run_retinanet_on_dataset(...)` that iterates over our image list and runs inference on each image;
- saving predictions into a unified format:  
  `(image_id, class_id, score, x1, y1, x2, y2)`.

---

#### Minimal usage example

```python
import torch
from PIL import Image
from torchvision.models.detection import (
    retinanet_resnet50_fpn_v2,
    RetinaNet_ResNet50_FPN_V2_Weights,
)

# 1. Load COCO-pretrained weights and model
weights = RetinaNet_ResNet50_FPN_V2_Weights.DEFAULT
model = retinanet_resnet50_fpn_v2(weights=weights)
model.eval()

# 2. Preprocessing pipeline associated with these weights
preprocess = weights.transforms()

# 3. Load an RGB image (path relative to the project root)
img = Image.open("example.jpg").convert("RGB")
input_tensor = preprocess(img)
batch = [input_tensor]  # list[Tensor] – one tensor per image

# 4. Inference
with torch.no_grad():
    outputs = model(batch)

# 5. Extract predictions for the first image in the batch
boxes = outputs[0]["boxes"]   # [N, 4]
labels = outputs[0]["labels"] # [N]
scores = outputs[0]["scores"] # [N]

# Optional: filter by score threshold
score_thresh = 0.5
mask = scores >= score_thresh
boxes = boxes[mask]
labels = labels[mask]
scores = scores[mask]

# Optional: map label IDs to COCO class names
categories = weights.meta["categories"]
for i in range(min(5, len(boxes))):
    cls_id = labels[i].item()
    cls_name = categories[cls_id]
    print(
        f"{i}: class={cls_name} (id={cls_id}), "
        f"score={scores[i].item():.3f}, box={boxes[i].tolist()}"
    )
```

### FCOS (single-shot, anchor-free) from `torchvision`

**Summary**  
We can use `torchvision`’s `fcos_resnet50_fpn` as our **single-shot, anchor-free object detector**, pre-trained on **MS COCO**.  
This model will complement RetinaNet: both are single-stage, but FCOS is **anchor-free**, so the comparison will highlight differences between anchor-based and anchor-free approaches.

---

#### Why FCOS?

- **Single-stage (single-shot)**: detects objects in one pass, similar to RetinaNet, so it fits our efficiency-oriented setup.
- **Anchor-free**: does not rely on predefined anchor boxes; bounding boxes are predicted directly from feature map locations.
- **Pre-trained on COCO**: ready-to-use weights, no need to train from scratch.
- **Official `torchvision` implementation**: stable API, very similar to Faster R-CNN and RetinaNet, easy to plug into our existing pipeline.

---

#### Documentation / References

**Official FCOS documentation in torchvision**  
- https://docs.pytorch.org/vision/main/models/fcos.html  
- https://docs.pytorch.org/vision/main/models/generated/torchvision.models.detection.fcos_resnet50_fpn.html  

**Original FCOS paper (anchor-free one-stage detector)**  
- FCOS: Fully Convolutional One-Stage Object Detection – https://arxiv.org/abs/1904.01355  

**Intro / tutorials on FCOS and anchor-free detection**  
- Anchor-free object detection with FCOS (inference tutorial):  
  https://debuggercafe.com/anchor-free-object-detection-inference-using-fcos-fully-connected-one-stage-object-detection/  
- FCOS and anchor-free detection explained:  
  https://learnopencv.com/fcos-anchor-free-object-detection-explained/  

---

#### Minimal usage example

```python
import torch
from PIL import Image
from torchvision.models.detection import (
    fcos_resnet50_fpn,
    FCOS_ResNet50_FPN_Weights,
)

# 1. Load COCO-pretrained weights and model
weights = FCOS_ResNet50_FPN_Weights.DEFAULT
model = fcos_resnet50_fpn(weights=weights)
model.eval()

# 2. Preprocessing pipeline associated with these weights
preprocess = weights.transforms()

# 3. Load an RGB image
img = Image.open("example.jpg").convert("RGB")
input_tensor = preprocess(img)
batch = [input_tensor]  # list[Tensor] – one tensor per image

# 4. Inference
with torch.no_grad():
    outputs = model(batch)

# 5. Extract predictions for the first image in the batch
boxes = outputs[0]["boxes"]   # [N, 4]
labels = outputs[0]["labels"] # [N]
scores = outputs[0]["scores"] # [N]

# Filter by confidence threshold
score_thresh = 0.5
mask = scores >= score_thresh
boxes = boxes[mask]
labels = labels[mask]
scores = scores[mask]

# Optional: map label IDs to COCO class names
categories = weights.meta["categories"]

print("Detections:", len(boxes))
for i in range(min(5, len(boxes))):
    cls_id = labels[i].item()
    cls_name = categories[cls_id]
    print(
        f"{i}: class={cls_name} (id={cls_id}), "
        f"score={scores[i].item():.3f}, box={boxes[i].tolist()}"
    )
```

**Comment by iaroslav-tkachenko** [comment]
### RetinaNet (single-shot, anchor-based)
I locally tested the setup on a random city photo (`example.jpg`). Using the RetinaNet test script, the model successfully detected several objects (mainly cars and pedestrians) with high confidence scores and printed their class IDs, scores, and bounding boxes to the console, confirming that inference works end-to-end.

**Comment by iaroslav-tkachenko** [comment]
### FCOS (single-shot, anchor-free)

I verified the FCOS setup by running the `fcos_resnet50_fpn` model (with COCO-pretrained weights) on the same random city photo we used for RetinaNet. The script successfully produced 12 detections (mainly cars and one pedestrian) with reasonable confidence scores and bounding boxes, which confirms that the anchor-free single-shot detector is working end-to-end on local setup.

**Comment by iaroslav-tkachenko** [comment]
### Backup option
if we ever need another single-shot, anchor-based model, we could also use ssd300_vgg16 from torchvision. It’s a classic early one-stage anchor-based detector, available in the same API as Faster R-CNN / RetinaNet.  Easy to plug in, and it works nicely as a “historical” baseline against more modern models.

But I personally think RetinaNet is already enough as our main model, because it’s stronger and well-documented. Adding more anchor-based models probably won’t give us much extra insight.
